src/pluggy/hooks.py

```diff
@@ -163,22 +151,41 @@ def varnames(func):
         except Exception:
             return ()
 
-    try:  # func MUST be a function or method here or we won't parse any args
-        spec = _getargspec(func)
-    except TypeError:
-        return (), ()
+    # func MUST be a function or method here or we won't parse any args
+    # this is what we return if it is not
+    non_function = (), ()
 
-    args, defaults = tuple(spec.args), spec.defaults
-    if defaults:
-        index = -len(defaults)
-        args, defaults = args[:index], tuple(args[index:])
+    if hasattr(inspect, "signature"):
+        try:
+            signature = inspect.signature(func)
+        except ValueError:
+            return non_function
+
+        empty = inspect.Parameter.empty
+        variable = (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        parameters = signature.parameters.values()
+        parameters = tuple(p for p in parameters if p.kind not in variable)
+        args = tuple(p.name for p in parameters if p.default is empty)
+        defaults = tuple(p.name for p in parameters if p.default is not empty)
+        mangled_self = inspect.ismethod(func)
     else:
-        defaults = ()
+        try:
+            spec = inspect.getargspec(func)
+        except TypeError:
+            return non_function
+
+        args, defaults = tuple(spec.args), spec.defaults
+        if defaults:
+            index = -len(defaults)
+            args, defaults = args[:index], tuple(args[index:])
+        else:
+            defaults = ()
+        mangled_self = False
 
     # strip any implicit instance arg
     # pypy3 uses "obj" instead of "self" for default dunder methods
     implicit_names = ("self",) if not _PYPY3 else ("self", "obj")
-    if args:
+    if args and not mangled_self:
```

Comment:
Apparently, `signature` strips `self` for bound methods, but pluggy also strips it for unbound methods. The only way to preserve the previous behavior is to only additionally strip it if it wasn't already stripped. Hence the guard here.